source/Locales/Compactness/Properties.lagda

```diff
@@ -0,0 +1,716 @@
+
+   † = directify₂-is-equal-to-directify S is ⁻¹
+
+   q : (U ≤ (⋁[ 𝒪 X ] ⁅ S [ T [ x ] ] ∣ x ∶ index T ⁆)) holds
+   q = U                                          ≤⟨ p     ⟩
+       directify (𝒪 X) S [ is ]                   ＝⟨ †    ⟩ₚ
+       directify₂ S [ is ]                        ＝⟨ refl ⟩ₚ
+       ⋁[ 𝒪 X ] ⁅ S [ T [ x ] ] ∣ x ∶ index T ⁆   ■
+
+\end{code}
+
+It follows from this that `is-compact-open` implies `is-compact-open'`.
+
+\begin{code}
+
+ compact-open-implies-compact-open' : (U : ⟨ 𝒪 X ⟩)
+                                    → is-compact-open  X U holds
+                                    → is-compact-open' X U holds
+ compact-open-implies-compact-open' U κ S q = ∥∥-functor † (κ S↑ δ p)
+  where
+   open JoinNotation (join-of (𝒪 X))
+
+   Xₚ = poset-of (𝒪 X)
+
+   S↑ : Fam 𝓦 ⟨ 𝒪 X ⟩
+   S↑ = directify (𝒪 X) S
+
+   δ : is-directed (𝒪 X) (directify (𝒪 X) S) holds
+   δ = directify-is-directed (𝒪 X) S
+
+   p : (U ≤[ Xₚ ] (⋁[ 𝒪 X ] S↑)) holds
+   p = U             ≤⟨ Ⅰ ⟩
+       ⋁[ 𝒪 X ] S    ＝⟨ Ⅱ ⟩ₚ
+       ⋁[ 𝒪 X ] S↑   ■
+        where
+         Ⅰ = q
+         Ⅱ = directify-preserves-joins (𝒪 X) S
+
+   † : (Σ is ꞉ index S↑ , (U ≤[ Xₚ ] S↑ [ is ]) holds)
+     → Σ (J , β) ꞉ SubFam S ,
+        is-Kuratowski-finite J × (U ≤[ Xₚ ] (⋁⟨ j ∶ J ⟩ S [ β j ])) holds
+   † = uncurry (finite-subcover-through-directification U S)
+
+\end{code}
+
+We now prove the converse which is a bit more difficult. We start with some
+preparation.
+
+Given a subset `P : ⟨ 𝒪 X ⟩ → Ω` and a family `S : Fam 𝓤 ⟨ 𝒪 X ⟩`, the type
+`Upper-Bound-Data P S` is the type of indices `i` of `S` such that `S [ i ]` is
+an upper bound of the subset `P`.
+
+\begin{code}
+
+module Characterization-Of-Compactness₂ (X : Locale (𝓤 ⁺) 𝓤 𝓤) where
+
+ open Some-Lemmas-On-Directification (𝒪 X)
+ open PosetNotation (poset-of (𝒪 X))
+ open PosetReasoning (poset-of (𝒪 X))
+ open Joins (λ x y → x ≤ y)
+
+ Upper-Bound-Data : 𝓟 {𝓣} ⟨ 𝒪 X ⟩ → Fam 𝓤 ⟨ 𝒪 X ⟩ → 𝓤 ⁺ ⊔ 𝓣  ̇
+ Upper-Bound-Data P S =
+  Σ i ꞉ index S , (Ɐ x ꞉ ⟨ 𝒪 X ⟩ , P x ⇒ x ≤ S [ i ]) holds
+
+\end{code}
+
+We now define the truncated version of this which we denote `has-upper-bound-in`:
+
+\begin{code}
+
+ has-upper-bound-in :  𝓟 {𝓣} ⟨ 𝒪 X ⟩ → Fam 𝓤 ⟨ 𝒪 X ⟩ → Ω (𝓤 ⁺ ⊔ 𝓣)
+ has-upper-bound-in P S = ∥ Upper-Bound-Data P S ∥Ω
+
+\end{code}
+
+Given a family `S`, we denote by `χ∙ S` the subset corresponding to the image of
+the family.
+
+\begin{code}
+
+ χ∙ : Fam 𝓤 ⟨ 𝒪 X ⟩ → ⟨ 𝒪 X ⟩ → Ω (𝓤 ⁺)
+ χ∙ S U = U ∈image (S [_]) , being-in-the-image-is-prop U (S [_])
+  where
+   open Equality carrier-of-[ poset-of (𝒪 X) ]-is-set
+
+\end{code}
+
+Given a Kuratowski-finite family `S`, the subset `χ∙ S` is a Kuratowski-finite
+subset.
+
+\begin{code}
+
+ χ∙-of-Kuratowski-finite-subset-is-Kuratowski-finite
+  : (S : Fam 𝓤 ⟨ 𝒪 X ⟩)
+  → is-Kuratowski-finite (index S)
+  → is-Kuratowski-finite-subset (χ∙ S)
+ χ∙-of-Kuratowski-finite-subset-is-Kuratowski-finite S 𝕗 = ∥∥-functor † 𝕗
+  where
```

Comment:
This should have a more conceptual (and maybe shorter?) proof: `𝕋 (χ∙ S)` should be the image of `S` and the map `index S → image S` is always a surjection, so you could simply compose the surjections, I think.